impl.go

```diff
@@ -305,6 +320,44 @@ func validReceiver(recv string) bool {
 	return err == nil
 }
 
+// commentsBefore returns true if comments precede the field. It returns false
+// otherwise.
+func commentsBefore(field *ast.Field, cg []*ast.CommentGroup) bool {
+	if len(cg) > 0 {
+		return cg[0].Pos() < field.Pos()
+	}
+	return false
+}
+
+// flattenCommentMap flattens the comment map to a string.
+// This function must be used at the point when m is expected to have a single
+// element.
+func flattenCommentMap(m ast.CommentMap) string {
+	if len(m) != 1 {
+		panic("flattenCommentMap expects comment map of length 1")
+	}
+	var result strings.Builder
+	for _, cgs := range m {
+		for _, cg := range cgs {
+			for _, c := range cg.List {
+				result.WriteString(c.Text)
+				// add an end-of-line character if this is '//'-style comment
+				if c.Text[1] == '/' {
+					result.WriteString("\n")
+				}
+			}
+		}
+	}
+
+	// for '/*'-style comments, make sure to append EOL character to the comment
```

Comment:
Strictly speaking, this is probably not correct:

```go
type x interface {
  /* who would do this? */ F()
}
```

but I'm not inclined to worry about it. :)

Reply:
Actually, I wonder whether it would be better instead to do:

```go
if s := result.String(); !s.HasSuffix("\n") {
  result.WriteString("\n")
}
```

What do you think?

Reply:
> Strictly speaking, this is probably not correct:
> 
> ```go
> type x interface {
>   /* who would do this? */ F()
> }
> ```
> 
> but I'm not inclined to worry about it. :)

Yeah I agree, nobody would ever want to do that, This check and EOL character is required though because `ast.CommentMap` would not have a EOL character at the end for /*-style comments, iirc. 

I was initially thinking about placing something like this
```
return strings.TrimSuffix(result.String(), "\n")+"\n" 
```
just to make sure there is always `\n` character at the end of the comments, but your code works just as good :-).